src/paint.rs

```diff
@@ -522,9 +522,15 @@ impl<'p> Painter<'p> {
             let line_has_emph_and_non_emph_sections =
                 style_sections_contain_more_than_one_style(style_sections);
             let should_update_non_emph_styles = non_emph_style.is_some() && *line_has_homolog;
-            let is_whitespace_error =
-                whitespace_error_style.is_some() && is_whitespace_error(style_sections);
-            for (style, _) in style_sections.iter_mut() {
+
+            // TODO: Git recognizes blank lines at end of file (blank-at-eof)
+            // as a whitespace error but delta does not yet.
+            // https://git-scm.com/docs/git-config#Documentation/git-config.txt-corewhitespace
+            let mut is_whitespace_error = whitespace_error_style.is_some();
+            for (style, s) in style_sections.iter_mut().rev() {
+                if is_whitespace_error && !s.trim().is_empty() {
+                    is_whitespace_error = false;
+                }
```

Comment:
Why does `!s.trim().is_empty()` mean that `s` is not a whitespace error? Isn't `s = "xxx  "` a counterexample? Are you redefining `is_whitespace_error` to mean "whole line is whitespace"?

Reply:
Because of the modification inside `edit.rs` I should not have to deal with `"xxx "` as a token but `["xxx", " "]` instead.
So the whole  `" "`is a token with whitespace error.

I am indeed re-using the `is_whitespace_error` to be flexible between "whole line is whitespace" and "part of the line is whitespace".

So to answer your question, since I am iterative with a `rev()`, I can say something like while I am encountering only whitespace, I apply the whitespace_error style. Once I encounter a non-whitespace token, I must stop applying the whitespace error style for the whole line

----

I also have a local change to consider this as an error :  `"   \t"` as asked by martinetd,  but it only works with the option `--tabs=0` so I do not want to bloat this PR with it. making it works with and without the tab expansion will make it a heavy refactor of the codebase (and I do not use tabs so ... :shrug: )

Reply:
> I also have a local change to consider this as an error : "   \t" as asked by martinetd, but it only works with the option --tabs=0 so I do not want to bloat this PR with it. making it works with and without the tab expansion will make it a heavy refactor of the codebase (and I do not use tabs so ... shrug )

Yeah; I think that can come later (if at all).

Unfortunately even if I do have a properly configured editor, patches I review often don't so displaying such problems can be useful.
In that case, just displaying "<space><tab>" as an error isn't enough as having someone add a line with tabs in the middle of a space-indented file is also a problem, so displaying tabs differently would also be necessary -- and we're getting quite out of scope of this PR.
(expanding the offtopic point: gerrit displays them with a lightly shaded `»   ` of appropriate size, which I kind of like and have reproduced in gertty at some point, so I might do that here eventually)

Your work so far is already more than useful, and it's great to see interest to merge it now. Thanks a bunch everyone! :)